File: problems/1/TwoSum.scala
object TwoSum {
    
    object Solution {
        def twoSum(nums: Array[Int], target: Int): Array[Int] = {
            var numMap: Map[Int, Int] = Map()
            for (i <- 0 until nums.length) {
                val targetNum = target - nums(i)
                if (numMap.contains(targetNum)) {
                    return Array(numMap(targetNum), i)
                }

                numMap += nums(i) -> i
            }
            return null
        }
    }

    def main(args: Array[String]) = {
        val res = Solution.twoSum(Array(2, 7, 11, 15), 9)
        println(res(0), res(1))
    }
}

File: problems/1/TwoSum.cc
#include <iostream>
#include <cstdio>
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, int> numMap;
        vector<int> res;
        for (int i = 0; i < nums.size(); i++) {
            const int targetNum = target - nums[i];
            if (numMap.find(targetNum) != numMap.end()) {
                res.push_back(numMap[targetNum]);
                res.push_back(i);
                return res;
            }

            numMap[nums[i]] = i;
        }

        return res;
    }
};

int main() {
    Solution solution;
    int a[] = {2, 7, 11, 15} ;
    vector<int> inp(a, a + 4);
    const vector<int> res = solution.twoSum(inp, 9);
    cout << res[0] << " " << res[1] << endl;
    return 0;
}
